ci.ocp: Improve the service-up detection

ldoktor opened the pull request:
waiting for the first response is not sufficient as OCP returns html page without error even when the route is not yet established describing the issue (why it doesn't reply with 500?). Waiting for the correct output should do better.

ldoktor commented:
I tested (similar) version here: https://github.com/openshift/release/pull/48473 seems to work well.

ldoktor commented:
@wainersm @gkurz could you please review this? It significantly decreases the stability of the CI pipeline.

stevenhorsman commented:
Hi Lukas - is there a plan to move this openshift testing into the kata-containers repo as the tests repo is deprecated?

stevenhorsman reviewed (commented):
I think we'll need an issue linked in the commit body via `Fixes:` to satisfy the automation

ldoktor commented on .ci/openshift-ci/run_smoke_test.sh:
```diff
@@ -57,14 +57,19 @@ else
-grep "${hello_msg}" hello_msg.txt > /dev/null
-test_status=$?
-if [ $test_status -eq 0 ]; then
+if waitForProcess 60 1 "curl '${host}:${port}${hello_file}' 2>/dev/null | grep -q '$hello_msg'"; then
```
The difference is described in commit message. Previously we waited for any successful curl reply which, quite often, is actually an error message from OCP saying the service is not yet ready (but still returns OK 200). While this version waits for the output to be the expected string so even successful curl reply that does not contains such message will be ignored and retried later.

gkurz replied:
Ok, I missed the double-quotes around the `curl | grep` pipeline... 

An option to avoid confusion and make this easier to review would be to do something like :
```
check_command="curl '${host}:${port}${hello_file}' 2>/dev/null | grep -q '$hello_msg'"
if waitForProcess 60 1 "$check_command"; then
```

ldoktor replied:
Sure, I can change that in v2

ldoktor commented on .ci/openshift-ci/run_smoke_test.sh:
```diff
@@ -57,14 +57,19 @@ else
 else
+	test_status=1
+	echo "::error:: HTTP server not working"
+	curl -vvv "${host}:${port}${hello_file}"
```
Well I can catch the full output of the `waitForProcess` and print it only on failure, but didn't seemed that important.

ldoktor replied:
Well to get the full curl output we'd have to `tee` it into a file as we need to grep for the output. I don't think it's worth the effort, what do you think?

tbuskey replied:
Capturing the output for a possible error message in the log is usually a good thing.
When things break, it helps having it in the log for debug.

ldoktor replied:
Sure, I know, let me add it to demonstrate the complexity....

ldoktor commented on .ci/openshift-ci/run_smoke_test.sh:
```diff
@@ -57,14 +57,19 @@ else
 	info "HTTP server is working"
 else
+	test_status=1
+	echo "::error:: HTTP server not working"
```
This file is consumed by prow job and (unless I'm mistaken) the logging formatters work there too.

ldoktor commented on .ci/openshift-ci/run_smoke_test.sh:
```diff
@@ -57,14 +57,19 @@ else
-if [ $test_status -eq 0 ]; then
+if waitForProcess 60 1 "curl '${host}:${port}${hello_file}' 2>/dev/null | grep -q '$hello_msg'"; then
+	test_status=0
 	info "HTTP server is working"
```
Well, this message was already there (and is quite useful when looking for things that do not work, at least you know this thing worked...). Anyway I'd rather move a discussion about this message to another PR, if needed.

ldoktor commented:
Changes:

* Using "check_cmd" variable to emphasize the waitForProcess command
* Split into 2 commits (fix and extra debug output)
* Capture output of all curl attempts and log them in case of a failure
* Added "Fixes" to the first commit

Not changed:

* Kept the "HTTP server is working" message as it was there before this PR

tbuskey commented:
/lgtm

gkurz reviewed (approved):
LGTM. Thanks @ldoktor !

gkurz commented:
> Hi Lukas - is there a plan to move this openshift testing into the kata-containers repo as the tests repo is deprecated?

@stevenhorsman IIUC Lukas is using the test repo as a guinea pig. When everything works, the openshift testing will go the the kata-containers repo like the rest.

ldoktor commented:
> > Hi Lukas - is there a plan to move this openshift testing into the kata-containers repo as the tests repo is deprecated?
> 
> @stevenhorsman IIUC Lukas is using the test repo as a guinea pig. When everything works, the openshift testing will go the the kata-containers repo like the rest.

Most of the work is done, now I'm waiting for the 4.13 pipeline issue to be resolved so the https://github.com/openshift/release/pull/48435 can be merged (we actually think about merging it with the failure, depends on the reviewers). Anyway this patchset will have to be backported there, I'll take care of it once it's ready.

wainersm reviewed (approved):
Good catch! Thanks @ldoktor !
